### PR TITLE
Allow for Thread reuse for CompactIndex fetcher

### PR DESCRIPTION
**Pull request by NickLaMuro**
Overview
--------
One of the main benefits of the `Net::HTTP::Persistent` library is that it maintains the HTTP connections so that multiple requests to the same host don't need to re-instantiate the http tcp socket to make a request, and also allows this to be done across threads.  But when used with the `CompactIndex` client, the each iteration of the `remaining_gems?` loop throws away the `Bundle::Worker` and initiates new Threads with new connections.

Since the resulting work on these threads only returns to the caller the result, and can be collected cleanly, it is safe to reuse these threads by other repeated `gem_name` iterations and gain a bit of a performance boost in the process.

While in this code change, we do run `@bundle_worker.stop`, we could simple skip this now as we will never use more then one `Bundle::Worker`'s worth of threads, keeping around the worker for further calls to `.specs` on the same fetcher.  That said, this has not been don at this time.

Metrics
-------
This is the result of running the fetcher for fetching gem metadata against the Gemfile for the [ManageIQ](https://github.com/ManageIQ/manageiq) project (these are sorted collection of results from the quickest times to the slowest times from a collection of benchmarks made over time, and are not in any other order besides that):

| No changes | Shared threads, 25 workers  | Shared threads, 15 workers |
| ---------- | --------------------------- | -------------------------- |
|  14.780272 |                    4.429732 |                   4.424576 |
|  14.987697 |                    4.664379 |                   4.447966 |
|  16.161742 |                    7.470335 |                   4.596416 |
|  16.698832 |                    8.538736 |                   8.103314 |
|  16.973982 |                   10.293958 |                  10.368032 |
|  17.044186 |                   12.548805 |                  10.873393 |
|  17.401267 |                   13.527048 |                  11.734078 |
|  17.417851 |                   13.640801 |                  12.379456 |
|  17.611628 |                   13.695791 |                  14.071921 |
|  17.702503 |                   13.930277 |                  14.151556 |
|  17.926312 |                   14.512490 |                  14.247501 |
|  17.953948 |                   14.547216 |                  14.372980 |
|  18.116374 |                   14.562247 |                  14.494275 |
|  18.193501 |                   15.054224 |                  14.496102 |
|  18.278697 |                   16.325938 |                  14.496330 |
|  18.313061 |                   16.773398 |                  14.799070 |
|  18.561052 |                   17.029987 |                  15.377202 |
|  18.657163 |                   17.321094 |                  15.658238 |
|  18.661411 |                   17.443478 |                  17.018813 |
|  18.739866 |                   17.895521 |                  17.142276 |
|  19.008614 |                   18.075015 |                  17.414206 |
|  19.174890 |                   18.159239 |                  17.860201 |
|  19.235642 |                   18.188842 |                  18.460573 |

The following changes to the `lib/bundler/source/rubygems.rb` were used to capture the metrics.

```diff
diff --git a/lib/bundler/source/rubygems.rb b/lib/bundler/source/rubygems.rb
index aedad70..2cd637b 100644
--- a/lib/bundler/source/rubygems.rb
+++ b/lib/bundler/source/rubygems.rb
@@ -350,11 +350,17 @@ module Bundler
             " Downloading full index instead..." unless allow_api

           if allow_api
+            require "benchmark"
+            puts Benchmark.measure {
             api_fetchers.each do |f|
               Bundler.ui.info "Fetching gem metadata from #{f.uri}", Bundler.ui.debug?
               idx.use f.specs_with_retry(dependency_names, self)
               Bundler.ui.info "" unless Bundler.ui.debug? # new line now that the dots are over
             end
+            }
+
+            puts; puts; puts "exiting..."
+            exit
```

And the command used to capture the metrics using this branch:

```
$ for I in `seq 1 15`; do \
    BUNDLE_DISABLE_POSTIT=1 ruby -I /path/to/bundler/lib /path/to/bundler/exe/bundle update; \
  done
```

Some things to note about this data:

- These results are the lowest of the recorded and usable results, not an average
- I went with the lowest because is show the range of improvement the best, but I didn't get an even number of usable results, so this was the best way (I found) to share a somewhat unbiased distribution
- Usable results were the following
  - Used the `compact_index` resolver
  - Didn't require a retry
- Unfortunately, almost half of my results had to be thrown away because they didn't match that criteria

One thing that I also observed while running in debug node and inspecting the threads is the it seems that one or two workers tend take a majority of the requests, while the other workers seem to only process one.  This is why the metrics show a second column where I adjusted the number of workers spawned to 15 (no reflected in this code change), something to also consider.

Tests
-----
Wasn't able to run tests locally when I using the latest version of master (not sure what broke), and unsure how I would test this functionality.  Let me know if this is something you would like to go forward with adding and I will look into adding some tests around it.

### CODE REVIEW

**Comment by indirect** [comment]
Oof. This is a great catch! Thanks for the PR.

**Comment by NickLaMuro** [comment]
No prob.  You and @segiddins have been on the ball with responding to my (cheeky) pull requests the past couple of days, and I appreciate the work you both are putting in.

I do want to emphasize with the metrics that they are a sorted sampling, and in no particular order aside from that.  The "super fast" ones happen, but not all the time, and are definitely dependent on the response times of the rubygems api and how well my wifi card wants to behave with my router...

Hopefully I can figure out what broke with the tests, but they were working just fine when I was running them before 😟

File: lib/bundler/fetcher/compact_index.rb [hunk 1]
```diff
@@ -44,6 +44,8 @@ def specs_for_names(gem_names)
           complete_gems.concat(deps.map(&:first)).uniq!
           remaining_gems = next_gems - complete_gems
         end
+        @bundle_worker.stop if @bundle_worker
+        @bundle_worker = nil # reset it, just incase.  Not sure if necessary
```
Comment (segiddins) [inline lib/bundler/fetcher/compact_index.rb]: remove `just` from here and it will pass :P

Comment (NickLaMuro) [inline lib/bundler/fetcher/compact_index.rb]: Okay, just found the test that explains that in `spec/quality_spec.rb` 👍 

To be fair, I was also getting test failures like this locally:

```
  1) bundle executable when ENV['BUNDLE_GEMFILE'] is set to an empty string ignores it
     Failure/Error: run! "require '#{name}.rb'; puts #{version_const}", *groups

     RuntimeError:
       Invoking run!("require 'rack.rb'; puts RACK", {}) failed:
       /Users/nicklamuro/code/public_forks/bundler/lib/bundler/resolver.rb:365:in `block in verify_gemfile_dependencies_are_found!': Could not find gem 'rack' in any of the gem sources listed in your Gemfile or available on this machine. (Bundler::GemNotFound)
        from /Users/nicklamuro/code/public_forks/bundler/lib/bundler/resolver.rb:340:in `each'
        from /Users/nicklamuro/code/public_forks/bundler/lib/bundler/resolver.rb:340:in `verify_gemfile_dependencies_are_found!'
        from /Users/nicklamuro/code/public_forks/bundler/lib/bundler/resolver.rb:198:in `start'
        from /Users/nicklamuro/code/public_forks/bundler/lib/bundler/resolver.rb:181:in `resolve'
        from /Users/nicklamuro/code/public_forks/bundler/lib/bundler/definition.rb:243:in `resolve'
        from /Users/nicklamuro/code/public_forks/bundler/lib/bundler/definition.rb:167:in `specs'
        from /Users/nicklamuro/code/public_forks/bundler/lib/bundler/definition.rb:226:in `specs_for'
        from /Users/nicklamuro/code/public_forks/bundler/lib/bundler/definition.rb:215:in `requested_specs'
        from /Users/nicklamuro/code/public_forks/bundler/lib/bundler/environment.rb:19:in `requested_specs'
        from /Users/nicklamuro/code/public_forks/bundler/lib/bundler/runtime.rb:14:in `setup'
        from /Users/nicklamuro/code/public_forks/bundler/lib/bundler.rb:100:in `setup'
        from -e:1:in `<main>'
     # ./spec/support/matchers.rb:60:in `block in should_be_installed'
     # ./spec/support/matchers.rb:57:in `each'
     # ./spec/support/matchers.rb:57:in `should_be_installed'
     # ./spec/bundler/cli_spec.rb:38:in `block (3 levels) in <top (required)>'
```

But maybe my computer was just as tired as I am... or was hallucinating...

**Comment by segiddins** [comment]
@indirect r?

**Comment by NickLaMuro** [comment]
@segiddins I accidentally pushed the worker count down to 15 after updating that last commit (after doing some testing), and also found out that my internet connection was running at an abysmal 1.8ish Mbs on my lappy for who knows how long (restart fixed it).  So I am going to re-push and add some updated metrics that are much more reasonable and consistent.

**Comment by NickLaMuro** [comment]
New (more reasonable) metrics using the ManangeIQ and OpenProject repos as test beds:

### ManageIQ  ( https://github.com/ManageIQ/manageiq )
| No changes | Shared threads, 25 workers  | Shared threads, 15 workers |
| ---------- | --------------------------- | -------------------------- |
|   1.421065 |                    1.225885 |                   1.235604 |
|   1.422998 |                    1.234367 |                   1.285666 |
|   1.448543 |                    1.240576 |                   1.294454 |
|   1.471257 |                    1.248634 |                   1.308538 |
|   1.472592 |                    1.256989 |                   1.308846 |
|   1.475861 |                    1.262683 |                   1.311133 |
|   1.479348 |                    1.264154 |                   1.315734 |
|   1.480403 |                    1.265855 |                   1.318963 |
|   1.482468 |                    1.281820 |                   1.322024 |
|   1.484387 |                    1.292216 |                   1.324417 |
|   1.490546 |                    1.310216 |                   1.325004 |
|   1.498497 |                    1.321688 |                   1.328751 |
|   1.512019 |                    1.322935 |                   1.334134 |
|   1.543380 |                    1.324550 |                   1.344470 |
|   1.567356 |                    1.338790 |                   1.349862 |
|   1.619723 |                    1.395702 |                   1.362866 |



### OpenProject  ( https://github.com/opf/openproject )
| No changes | Shared threads, 25 workers  | Shared threads, 15 workers |
| ---------- | --------------------------- | -------------------------- |
|   1.770085 |                    1.331694 |                   1.405903 |
|   1.798816 |                    1.335906 |                   1.413870 |
|   1.799796 |                    1.345552 |                   1.414583 |
|   1.802692 |                    1.407406 |                   1.414614 |
|   1.807121 |                    1.408128 |                   1.417509 |
|   1.813796 |                    1.409145 |                   1.418434 |
|   1.822499 |                    1.422775 |                   1.420102 |
|   1.831169 |                    1.425031 |                   1.423495 |
|   1.832604 |                    1.434294 |                   1.430579 |
|   1.840357 |                    1.435858 |                   1.433573 |
|   1.845169 |                    1.440178 |                   1.436723 |
|   1.889174 |                    1.441677 |                   1.439732 |
|   1.902441 |                    1.444408 |                   1.440461 |
|   1.919368 |                    1.444862 |                   1.440594 |
|   2.186560 |                    1.466290 |                   1.468714 |
|   2.338394 |                    1.467803 |                   1.490589 |

**Comment by indirect** [comment]
@homu r+

**Pull request merged**